### PR TITLE
Fix Issue 21014 - aa.byKeyValue, byKey, byValue very under-documented

### PR DESCRIPTION
**Pull request by wolframw**
This PR aims to improve the documentation and the examples of the AA-specific functions `byKeyValue`, `byKey`, `byValue`, `keys`, and `values`.

### CODE REVIEW

File: src/object.d [hunk 1]
```diff
@@ -2771,11 +2771,28 @@ private AARange _aaToRange(T: V[K], K, V)(ref T aa) pure nothrow @nogc @safe
...
 
 /***********************************
- * Returns a forward range over the keys of the associative array.
+ * Returns a $(REF_ALTTEXT forward range, isForwardRange, std,range,primitives)
+ * which will iterate over the keys of the associative array. The keys are
+ * returned by reference.
+ *
+ * If changes are made to the associative array, all ranges previously obtained
+ * through this function are invalidated. The following example program will
+ * crash:
```
Comment (schveiguy) [inline src/object.d:2780]: Not exactly true. If you change just a value, the iteration will be fine.

Maybe say "If structural changes are made to the associative array (removing or adding keys), all ranges..."

This applies to all the other locations where you have this message.

I would also change "crash" to "dereference a null pointer" to clarify the behavior is not unsafe.

File: src/object.d [hunk 2]
```diff
@@ -2807,20 +2824,37 @@ auto byKey(T : V[K], K, V)(T* aa) pure nothrow @nogc
 ///
 @safe unittest
 {
-    auto dict = [1: 0, 2: 0];
-    int sum;
+    auto dict = [1: "v1", 2: "v2"];
+    int sum = 0;
```
Comment (schveiguy) [inline src/object.d:2828]: Unclear why this change is made.

File: src/object.d [hunk 3]
```diff
@@ -2914,21 +2973,25 @@ auto byKeyValue(T : V[K], K, V)(T* aa) pure nothrow @nogc
...
 @safe unittest
 {
+    import std.stdio : writeln;
+
     auto dict = ["k1": 1, "k2": 2];
-    int sum;
+    int sum = 0;
     foreach (e; dict.byKeyValue)
+    {
+        writeln(e.key, ": ", e.value);
```
Comment (schveiguy) [inline src/object.d:2982]: No writeln in unittests. Use assert. It will replace the call with writeln in the docs.

Maybe `assert(e.key[1] == e.value + '0')`

File: src/object.d [hunk 4]
```diff
@@ -2914,21 +2973,25 @@ auto byKeyValue(T : V[K], K, V)(T* aa) pure nothrow @nogc
...
         sum += e.value;
+    }
 
     assert(sum == 3);
 }
 
 /***********************************
- * Returns a dynamic array, the elements of which are the keys in the
- * associative array.
+ * Copies the keys of the associative array into a dynamic array and returns it.
```
Comment (schveiguy) [inline src/object.d:2990]: A little awkward wording. Maybe:

"returns a newly allocated dynamic array containing a copy of the keys from the associative array"

File: src/object.d [hunk 5]
```diff
@@ -2975,12 +3038,12 @@ Key[] keys(T : Value[Key], Value, Key)(T *aa) @property
 }
 
 /***********************************
- * Returns a dynamic array, the elements of which are the values in the
- * associative array.
+ * Copies the values of the associative array into a dynamic array and returns
+ * it.
```
Comment (schveiguy) [inline src/object.d:3042]: Similar treatment.

**Review by schveiguy — CHANGES_REQUESTED** [review]
Overall looks pretty good. Thanks for tackling this!

**Comment by schveiguy** [comment]
FYI the reason the CI is failing is due to the import of std.stdio in the unittests.

File: src/object.d (hunk 2 above)
Comment (wolframw) [reply src/object.d:2828]: In the other examples, the keys and values are of different types and it's the values being added up. When I looked at this example, it honestly took me a bit of time to figure out that it's the keys being added up, so I thought having different types for keys and values might be an improvement.

About the initialization of `sum`: I was under the impression that using uninitialized variables was a bad practice, even in D land. I was wrong.

File: src/object.d [hunk 6]
```diff
@@ -2771,11 +2771,27 @@ private AARange _aaToRange(T: V[K], K, V)(ref T aa) pure nothrow @nogc @safe
...
+ * ranges previously obtained through this function are invalidated. The
+ * following example program will dereference a null pointer:
+ *
+ *---
+ * import std.stdio : writeln;
+ *
+ * auto dict = ["k1": 1, "k2": 2];
+ * auto keyRange = dict.byKey;
+ * dict.clear;
+ * writeln(keyRange.front);
```
Comment (RazvanN7) [inline src/object.d:2788]: Maybe add a comment to this line that says "Segmentation fault"

File: src/object.d [hunk 7]
```diff
@@ -2771,11 +2771,27 @@ private AARange _aaToRange(T: V[K], K, V)(ref T aa) pure nothrow @nogc @safe
...
+ *---
+ *
  * Params:
  *      aa =     The associative array.
  * Returns:
- *      A forward range.
+ *      A forward range over the keys of the associative array.
```
Comment (RazvanN7) [inline src/object.d:2794]:
```suggestion
 *      A forward range containing the keys of the associative array.
```

File: src/object.d [hunk 8]
```diff
@@ -2816,11 +2832,27 @@ auto byKey(T : V[K], K, V)(T* aa) pure nothrow @nogc
...
+ * ranges previously obtained through this function are invalidated. The
+ * following example program will dereference a null pointer:
+ *
+ *---
+ * import std.stdio : writeln;
+ *
+ * auto dict = ["k1": 1, "k2": 2];
+ * auto valueRange = dict.byValue;
+ * dict.clear;
+ * writeln(valueRange.front);
```
Comment (RazvanN7) [inline src/object.d:2849]: ditto

File: src/object.d [hunk 9]
```diff
@@ -2816,11 +2832,27 @@ auto byKey(T : V[K], K, V)(T* aa) pure nothrow @nogc
...
+ *---
+ *
  * Params:
  *      aa =     The associative array.
  * Returns:
- *      A forward range.
+ *      A forward range over the values of the associative array.
```
Comment (RazvanN7) [inline src/object.d:2855]: Same as above

File: src/object.d [hunk 10]
```diff
@@ -2861,11 +2893,35 @@ auto byValue(T : V[K], K, V)(T* aa) pure nothrow @nogc
...
+ * ranges previously obtained through this function are invalidated. The
+ * following example program will dereference a null pointer:
+ *
+ *---
+ * import std.stdio : writeln;
+ *
+ * auto dict = ["k1": 1, "k2": 2];
+ * auto kvRange = dict.byKeyValue;
+ * dict.clear;
+ * writeln(kvRange.front.key, ": ", kvRange.front.value);
```
Comment (RazvanN7) [inline src/object.d:2912]: I think that this is a bit misleading. The segfault will occur because of `front`. Does it really matter if we access `key` or `value`?

File: src/object.d [hunk 11]
```diff
@@ -2861,11 +2893,35 @@ auto byValue(T : V[K], K, V)(T* aa) pure nothrow @nogc
...
+ * array and is not compatible withth the
+ * $(LINK2 $(ROOT_DIR)phobos/std_typecons.html#.Tuple,`Tuple`) type in Phobos.
+ * For compatibility with `Tuple`, use
+ * $(LINK2 $(ROOT_DIR)phobos/std_array.html#.byPair,std.array.byPair) instead.
+ *
  * Params:
  *      aa =     The associative array.
  * Returns:
- *      A forward range.
+ *      A forward range over the pairs of the associative array.
```
Comment (RazvanN7) [inline src/object.d:2924]:
```suggestion
 *      A forward range containing the pairs of the associative array.
```

Comment (RazvanN7) [reply src/object.d:2924]: Or maybe it would be better to say "A forward range referencing ..."

Comment (RazvanN7) [reply src/object.d:2924]: This goes for all of the above and below similar changes

**Comment by RazvanN7** [comment]
@schveiguy Are we good with this PR?

**Comment by RazvanN7** [comment]
ping @wolframw

**Comment by RazvanN7** [comment]
@wolframw are you still pursuing this PR?

**Comment by RazvanN7** [comment]
@wolframw Please apply the requested changes so that we can merge this PR

**Comment by RazvanN7** [comment]
Closing due to inactivity. @wolframw feel free to reopen and apply the changes.

**Comment by wolframw** [comment]
> Closing due to inactivity. @wolframw feel free to reopen and apply the changes.

Sorry for the inactivity. I'd like to continue with this (in fact, it's already committed and pushed to my fork) but I don't have the rights to reopen this PR. Could you do that for me, please?

File: src/object.d (hunk 10 above)
Comment (wolframw) [reply src/object.d:2912]: `kvRange.front` will return `Pair(null, null)`, so I explicitly access `key` and `value` to make it blow up.

**Comment by thewilsonator** [comment]
> Could you do that for me, please?

No, it says  "the `fix_21014`branch was force-pushed or recreated". 

Just open another PR and link to this one.
